core/embed/rust/src/translations/blob.rs

```diff
@@ -0,0 +1,381 @@
+use core::{mem, str};
+
+use crate::{
+    crypto::{cosi, ed25519, merkle::merkle_root, sha256},
+    error::Error,
+    io::InputStream,
+};
+
+use super::public_keys;
+
+pub const MAX_HEADER_LEN: u16 = 1024;
+pub const EMPTY_BYTE: u8 = 0xFF;
+const SENTINEL_ID: u16 = 0xFFFF;
+
+const SIGNATURE_THRESHOLD: u8 = 2;
+
+// Maximum padding at the end of an offsets table (typically for alignment
+// purposes). We allow at most 3 for alignment 4. In practice right now this
+// should be max 1.
+const MAX_TABLE_PADDING: usize = 3;
+
+const INVALID_TRANSLATIONS_BLOB: Error = value_error!("Invalid translations blob");
+
+#[repr(packed)]
+struct OffsetEntry {
+    pub id: u16,
+    pub offset: u16,
+}
+
+pub struct Table<'a> {
+    offsets: &'a [OffsetEntry],
+    data: &'a [u8],
+}
+
+fn validate_offset_table(
+    data_len: usize,
+    mut iter: impl Iterator<Item = u16>,
+) -> Result<(), Error> {
+    // every offset table must have at least the sentinel
+    let mut prev = iter.next().ok_or(INVALID_TRANSLATIONS_BLOB)?;
+    for next in iter {
+        // offsets must be in ascending order
+        if prev > next {
```

Comment:
we need to for the translation strings to work. we never remove an index (so that an older firmware looking at a newer blob sees the indices in the right places) so any removed strings will be stored as zero-length strings